python/rapidsmpf/rapidsmpf/integrations/dask/shuffler.py

```diff
@@ -22,24 +23,57 @@
 if TYPE_CHECKING:
     from collections.abc import Callable, MutableMapping, Sequence
 
-    from distributed import Worker
+    from distributed import Client, Worker
 
 
-_shuffle_counter: int = 0
+# Set of available shuffle IDs
+_shuffle_id_vacancy: set[int] = set(range(Shuffler.max_concurrent_shuffles))
+_shuffle_id_vacancy_lock: threading.Lock = threading.Lock()
 
 
-def get_shuffle_id() -> int:
+def _get_new_shuffle_id(client: Client) -> int:
     """
-    Return the unique id for a new shuffle.
+    Get a new available shuffle ID.
+
+    Since RapidsMPF only supports a limited number of shuffler instances at
+    any given time, this function maintains a shared pool of shuffle IDs.
+
+    If no IDs are available locally, it queries all workers for IDs in use,
+    updates the vacancy set accordingly, and retries. If all IDs are in use
+    across the cluster, an error is raised.
+
+    Parameters
+    ----------
+    client
+        A Dask distributed client used to query workers for active shuffle IDs.
 
     Returns
     -------
-    The enumerated integer id for the current shuffle.
+    A unique shuffle ID not currently in use.
+
+    Raises
+    ------
+    ValueError
+        If all shuffle IDs are currently in use across the cluster.
     """
-    global _shuffle_counter  # noqa: PLW0603
+    with _shuffle_id_vacancy_lock:
+        if not _shuffle_id_vacancy:
+
+            def get_vacancy_ids(dask_worker: Worker) -> set[int]:
+                ctx = get_worker_context(dask_worker)
+                with ctx.lock:
+                    return set(ctx.shufflers.keys())
+
+            _shuffle_id_vacancy.update(range(Shuffler.max_concurrent_shuffles))
```

Comment:
Is there a reason to use `update` instead of `_shuffle_id_vacancy = set(range(Shuffler.max_concurrent_shuffles))`? Not a big deal, it just took me a while to understand what was being done here (assuming my understanding is even correct that the two are equivalent).

Reply:
Only to avoid to declare `_shuffle_id_vacancy` global :)
Changed to:
```python
_shuffle_id_vacancy = set(range(Shuffler.max_concurrent_shuffles))
```

Reply:
I wasn't proposing making it global, I'm fine with either the old or new implementation, TBH I didn't realize that's what you were avoiding thus I asked because it didn't look the most natural way to do it.

Reply:
Sure, but I think using global make the code more clear, in this case.

Reply:
Sounds good to me. Thanks!